Output `class_hash` and `compiled_class_hash` when build cairo1

abulenok opened the pull request:
- [x] Add https://github.com/software-mansion/protostar/pull/1905#discussion_r1192154950

abulenok committed:
Log compiled_class_hash when running build-cairo1

abulenok committed:
Log class_hash when running build-cairo1

abulenok committed:
Add _prepare_build_cairo1_args to integration/ProtostarFixture

Arcticae commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -12,6 +24,24 @@
+    with open(sierra_contract_file_path, mode="r", encoding="utf-8") as file:
+        sierra_compiled = json.loads(file.read())
+        sierra_compiled.pop("sierra_program_debug_info", None)
+        sierra_compiled["abi"] = json.dumps(sierra_compiled["abi"])
```
Is ABI necessary for this to work?

Arcticae commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -67,15 +97,32 @@ async def _build_contract(
+            output_path=sierra_compiled_contract_path,
+        )
+        class_hash = compute_class_hash_from_path(sierra_compiled_contract_path)
+        logging.info(
```
the command should use `messenger` pattern, to also support json output

abulenok commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -12,6 +24,24 @@
+    with open(sierra_contract_file_path, mode="r", encoding="utf-8") as file:
+        sierra_compiled = json.loads(file.read())
+        sierra_compiled.pop("sierra_program_debug_info", None)
+        sierra_compiled["abi"] = json.dumps(sierra_compiled["abi"])
```
What is the reason to drop ABI here?

abulenok replied:
An error occurs, if the ABI is dropped

```
marshmallow.exceptions.ValidationError: {'abi': ['Missing data for required field.']}
```

abulenok committed:
Add --json to build-cairo1

abulenok committed:
Write compiled_class_hash and class_hash to the files when build-cairo1

abulenok committed:
Don't check full class hash in build-cairo1 tests

abulenok committed:
Update documentation for build-cairo1

abulenok committed:
Add missing text to build-cairo1 description

karol-bisztyga reviewed (commented):
I'm going to do a very high-level review for now. I think we should definitely try to reuse the logic from the `declare` hint local. 

Have you considered it/looked at it?

There's that function used there `starkware.starknet.core.os.contract_class.compiled_class_hash.compute_compiled_class_hash`, using it would be cool as we wouldn't have to write additional code for this.

I think a lot of stuff can be copied from `declare` at least (if that works, we should extract mutual logic somewhere so it can be reused in both places (here and in `declare`).

abulenok committed:
Add missing paths to expected_built_files in test cairo 0 1 mixed building

abulenok committed:
Extract make_class funcs to ProjectCompiler

abulenok committed:
Add compile_contract to ProjectCompiler

abulenok committed:
Fix output file naming in build-cairo1

abulenok committed:
Add integration test for build-cairo1

abulenok committed:
Remove duplicated asserts from build-cairo1

abulenok committed:
Remove unused functions from build-cairo1

abulenok committed:
Remove unused functions from ProjectCompiler

abulenok committed:
Add compile_contract to MockedProjectCompiler in declare_cairo1_command_test

abulenok committed:
Use newer version of cairo submodule

... and enhance tests for cairo_python_bindings

abulenok committed:
Fix mock in declare cairo1 command test

abulenok committed:
Add return type for ProjectCompiler compile_contract

karol-bisztyga commented on protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:
```diff
@@ -64,31 +67,16 @@ def declare(contract: int) -> DeclaredContract:
-            sierra_compiled = json.loads(sierra_compiled)
-            sierra_compiled.pop("sierra_program_debug_info", None)
-            sierra_compiled["abi"] = json.dumps(sierra_compiled["abi"])
+            contract_class = make_contract_class(sierra_compiled)
+            compiled_class = make_compiled_class(casm_compiled)
```
```suggestion
            return make_contract_class(sierra_compiled), make_compiled_class(casm_compiled)
```

karol-bisztyga commented on protostar/cheatable_starknet/callable_hint_locals/declare_hint_local.py:
```diff
@@ -64,31 +67,16 @@ def declare(contract: int) -> DeclaredContract:
-            )
-
-            return CompiledClass.loads(casm_compiled)
+            return compiled_class, contract_class
```
```suggestion

```

karol-bisztyga commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -1,30 +1,92 @@
+    def format_human(self, fmt: LogColorProvider) -> str:
+        lines: list[str] = [
+            "Building cairo1 contracts",
+            f'Class hash for contract "{self.contract_name}": {hex(int(self.class_hash))}',
+            f'Compiled class hash for contract "{self.contract_name}": {hex(int(self.compiled_class_hash))}',
```
why do you cast to `int` variables that are already `int`s?

karol-bisztyga commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -1,30 +1,92 @@
+    def format_dict(self) -> dict:
+        return {
+            self.contract_name: {
+                "class_hash": hex(int(self.class_hash)),
+                "compiled_class_hash": hex(int(self.compiled_class_hash)),
```
why do you cast to int variables that are already `int`s?

karol-bisztyga commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -36,59 +98,65 @@ def name(self) -> str:
     @property
     def description(self) -> str:
-        return "Compile cairo1 contracts."
+        return "Compile cairo1 contracts. Writes `class_hash` and `compiled_class_hash` to the files and to stdout."
```
I'm not sure that this is necessary tbh.

karol-bisztyga commented on protostar/compiler/project_compiler.py:
```diff
@@ -29,42 +48,55 @@ def __init__(
-        return self._compile_cairo1_contract_from_contract_name(
-            contract_name=contract_name, output_type=_OutputType.SIERRA
+    def compile_contract(
+        self,
+        contract_identifier: ContractIdentifier,
+        cairo_path: Optional[list[Path]] = None,
+        output_path: Optional[Path] = None,
```
these repetitive args look like some boilerplate, maybe we could replace them with a data class or something?

Arcticae replied:
i disagree on this one, not that many args here, it would make things less readable

abulenok replied:
I also don't think there are too many args here. Grouping them would be "artificial"

karol-bisztyga replied:
ok, so be it

karol-bisztyga commented on tests/integration/_conftest/protostar_fixture.py:
```diff
@@ -291,6 +291,12 @@ def _prepare_build_args(self):
         args.contract_name = ""
         return args
 
+    def _prepare_build_cairo1_args(self):
+        args = self._prepare_build_args()
+        delattr(args, "cairo_path")
+        args.linked_libraries = None
+        return args
```
can you explain this change a bit?

karol-bisztyga commented:
> As of this method:
> 
> https://github.com/software-mansion/protostar/blob/1ab98364d2bfab4681541fd0ac3ba78f0df58af9/protostar/compiler/project_compiler.py#L54-L66
> 
> it is better to compile to `casm` using `sierra_compiled` string.
> 
> But I think this should be added in a separate PR, along with the tests for the to-be-created function in `cairo_python_bindings`. This PR is already big enough and the proposed refactor is not directly related with `class_hash` and `compiled_class_hash`.

This looks like a tiny change really, maybe we could add it here?

Arcticae commented on tests/e2e/cairo1/test_cairo1_building.py:
```diff
@@ -9,14 +10,40 @@
+    os.chdir("./cairo_project")
+    result_json = protostar(["build-cairo1", "--json"])
+
+    output_json_parsed = json.loads(
+        result_json.split("\n")[
+            3 if "This project expects older Protostar" in result_json else 0
+        ]
+    )
```
Isn't this a bug? 
`This project expects older Protostar`
We shouldn't output this kind of info in plaintext if user expects ndjson on the output, right?

abulenok replied:
But how the user will know he is using the wrong version of Protostar? Using the wrong version could result in unexpected behavior

cptartur replied:
I think ndjson spec allows returning multiple jsons separated by newline character. Perhaps version warning should be returned as a separate json?

Arcticae replied:
It's out of scope probably even if it's true, but i thought it was worth noting. Perhaps we can create an issue for this kind of stuff? cc @abulenok

MaksymilianDemitraszek replied:
warnings are redirected to stderr, ndjson should only be used in stdout. There is no need to do so.

abulenok replied:
agree with @MaksymilianDemitraszek, for test I've added `ignore_stderr=True`

abulenok committed:
Simplify

Co-authored-by: karol-bisztyga <59833830+karol-bisztyga@users.noreply.github.com>

abulenok committed:
Simplify

Co-authored-by: karol-bisztyga <59833830+karol-bisztyga@users.noreply.github.com>

abulenok committed:
Apply suggestions from the review

cptartur commented on tests/integration/cairo1/compiler_bindings/starknet_contracts_compiler_test.py:
```diff
@@ -6,19 +6,27 @@
+    )
+    assert cairo1.compile_starknet_contract_to_casm_from_path(
+        input_path=datadir / "basic_starknet_project",
+        output_path=datadir / "basic_starknet_project.casm",
+    )
+    assert cairo1.compile_starknet_contract_to_casm_from_path(
+        input_path=datadir / "test_starknet_project",
+        output_path=datadir / "test_starknet_project.casm",
```
These tests don't really verify if output_path contains anything useful. Is that on purpose?

abulenok replied:
Before recent changes, cairo bindings didn't return anything, when `output_path` was provided. It has changed in recent PRs.
But I have improved these tests in the newest commits

cptartur commented on tests/integration/commands/declare_cairo1_command_test.py:
```diff
@@ -46,6 +46,16 @@ def compile_contract_to_casm_from_contract_name(
             assert compiled is not None
             return compiled
```
Are this and method above still used by the test? If not, parts of this mock could be removed.

abulenok committed:
Rm unused import

abulenok committed:
Update docs

abulenok commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -36,59 +98,65 @@ def name(self) -> str:
     @property
     def arguments(self):
         return [
+            *MessengerFactory.OUTPUT_ARGUMENTS,
```
It doesn't work like that. For instance, look at the `test-cairo1` command

abulenok commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -36,59 +98,65 @@ def name(self) -> str:
+        cairo_path = linked_libraries + maybe_fetch_linked_libraries_from_scarb(
+            package_root_path=contract_paths[0],
+            linked_libraries=linked_libraries,
         )
```
The same assert is available in ProjectCompiler _compile_cairo1_contract_from_contract_name

abulenok committed:
Simplify mock

abulenok committed:
Check if file exists in test_contract_to_casm

cptartur commented on protostar/commands/cairo1_commands/build_cairo1_command.py:
```diff
@@ -1,30 +1,92 @@
+    with open(output_path, mode="w", encoding="utf-8") as output_file:
+        output_file.write(f"{hex(int(compiled_class_hash))}")
+
+    return compiled_class_hash
```
I think we already have code similiar to that inside ContractsController for declaring contracts.

abulenok replied:
I've moved `make_compiled_class` and `make_contract_class` to the `contracts` module. But the `compute_compiled_class_hash_from_casm_code` and `compute_class_hash_from_sierra_code` functions are specific to `build` command, I don't see a need to extract them to a separate module for now

abulenok committed:
Update test_contract_to_sierra_to_casm

Check if compile_starknet_contract_to_sierra_from_path + compile_starknet_contract_sierra_to_casm_from_path == compile_starknet_contract_to_casm_from_path

Arcticae commented on protostar/compiler/project_compiler.py:
This comment seems stale, right?

cptartur replied:
I think it's still relevant

abulenok committed:
Add compile_starknet_contract_sierra_to_casm_from_sierra_code

abulenok committed:
ignore_stderr in test_cairo1_build_json

abulenok committed:
Use triple quotes instead of string join

abulenok committed:
Move make_contract_class and make_compiled_class to contacts module

abulenok committed:
Merge branch 'master' into output-compiled-class-hash-when-build-cairo1

abulenok committed:
Refactor build command after merging master

abulenok committed:
Fix adding suffix for output path in Cairo1ContractCompiler

abulenok committed:
Remove info about build-cairo1

abulenok committed:
Uncomment git pull in install_cairo_bindings

abulenok committed:
Removed ProjectCompiler

abulenok committed:
Update cli docs

abulenok committed:
Add info about --json flag to 01-compiling docs

abulenok committed:
Use sierra compiled code when compiling to casm

abulenok committed:
Merge branch 'master' into output-compiled-class-hash-when-build-cairo1

abulenok committed:
Fix cairo bindings import error

abulenok committed:
Use build instead of build-cairo1

Arcticae commented on protostar/cheatable_starknet/controllers/contracts.py:
```diff
@@ -46,6 +46,18 @@
 from protostar.starknet.data_transformer import CairoData
 
 
+def make_contract_class(sierra_compiled: str):
```
return type should be annotated

Arcticae commented on protostar/cheatable_starknet/controllers/contracts.py:
```diff
@@ -46,6 +46,18 @@
+    return ContractClass.load(sierra_compiled_dict)
+
+
+def make_compiled_class(casm_compiled: str):
```
return type should be annotated

Arcticae commented on protostar/commands/cairo1_commands/build_command.py:
```diff
@@ -1,31 +1,91 @@
-from protostar.protostar_exception import ProtostarException
+
+
+def compute_class_hash_from_sierra_code(sierra_compiled: str, output_path: Path):
+    contract_class = make_contract_class(sierra_compiled)
+    class_hash = compute_class_hash(contract_class)
+
+    with open(output_path, mode="w", encoding="utf-8") as output_file:
+        output_file.write(f"{hex(class_hash)}")
+
+    return class_hash
+
+
+def compute_compiled_class_hash_from_casm_code(casm_compiled: str, output_path: Path):
+    compiled_class = make_compiled_class(casm_compiled)
+    compiled_class_hash = compute_compiled_class_hash(compiled_class)
+
+    with open(output_path, mode="w", encoding="utf-8") as output_file:
+        output_file.write(f"{hex(compiled_class_hash)}")
+
+    return compiled_class_hash
```
Those along with `make_contract_class`, `make_compiled_class` could be in a separate module called `contract_class`

Arcticae reviewed (approved):
Some nitpicking, looks very good to me, feel free to merge after 2nd approve

abulenok committed:
Annotate return type for make_contract_class and make_compiled_class

abulenok committed:
Extract make_contract_class and compute_class_hash from code to separate module

cptartur commented on protostar/cairo/contract_class.py:
```diff
@@ -0,0 +1,43 @@
+    return CompiledClass.loads(casm_compiled)
+
+
+def compute_class_hash_from_sierra_code(sierra_compiled: str, output_path: Path):
+    contract_class = make_contract_class(sierra_compiled)
+    class_hash = compute_class_hash(contract_class)
+
+    with open(output_path, mode="w", encoding="utf-8") as output_file:
+        output_file.write(f"{hex(class_hash)}")
+
+    return class_hash
+
+
+def compute_compiled_class_hash_from_casm_code(casm_compiled: str, output_path: Path):
+    compiled_class = make_compiled_class(casm_compiled)
```
Return type annotations missing in these functions

cptartur commented on protostar/cairo/contract_class.py:
```diff
@@ -0,0 +1,43 @@
+    contract_class = make_contract_class(sierra_compiled)
+    class_hash = compute_class_hash(contract_class)
+
+    with open(output_path, mode="w", encoding="utf-8") as output_file:
+        output_file.write(f"{hex(class_hash)}")
```
I don't really like this approach of always writing to file when computing these. Write should be done separately in the methods that require it.

abulenok replied:
I think we can leave writing to `output_path` here, but make it optional. This way, it will behave similarly to functions from `cairo_bindings`.

cptartur replied:
That's better but I still would prefer if writing to file was moved entirely to the module needing it. 

I'm assuming writing output of this command to file is needed only in the single place of the library.

cptartur commented on protostar/compiler/cairo1_contract_compiler.py:
```diff
@@ -0,0 +1,110 @@
+        cairo_path: Optional[list[Path]] = None,
+        output_path: Optional[Path] = None,
+    ) -> str:
+        if output_path:
+            output_path = output_path.with_suffix(".sierra.json")
```
I'm not sure if we should be modifying the path provided.

abulenok replied:
I removed this and added `output_path` creation to `compile_contract`

cptartur commented on protostar/compiler/cairo1_contract_compiler.py:
```diff
@@ -0,0 +1,110 @@
+                contract_name=contract_name, err=ex
+            ) from ex
+
+        assert sierra_compiled is not None
```
I think the assert here is no longer neede here. The bindings are now typed to always return a value.

cptartur commented on tests/e2e/cairo1/test_cairo1_building.py:
```diff
@@ -9,14 +10,37 @@
     assert compiled_casm_path.exists()
     assert compiled_casm_path.read_text()
 
+    class_hash_path = Path("build/main.class.hash")
+    compiled_class_hash_path = Path("build/main.compiled.class.hash")
```
I'm not sure where exactly this change was introduced, but I don't think we should be using `compiled.class.hash`.

If we want to have a custom extension for these I'd change it to `.compiled_class_hash` and `.compiled_class` respectively.

abulenok committed:
Add return type annotations

abulenok committed:
Remove redundant asserts

abulenok committed:
Move addition of suffixes to output path to compile_contract

abulenok committed:
Pass output_dir to compile_contract instead of output_path

abulenok committed:
Rename .class.hash, .compiled.class.hash -> .class_hash, .compiled_class_hash

abulenok committed:
Move writing class hashes to _build_contract

abulenok committed:
Merge branch 'master' into output-compiled-class-hash-when-build-cairo1
